File: csharp/src/Ice/Internal/ObjectAdapterFactory.cs
// Copyright (c) ZeroC, Inc.

using System.Net.Security;

namespace Ice.Internal;
public sealed class ObjectAdapterFactory
{
    public void shutdown()
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            //
            // Ignore shutdown requests if the object adapter factory has
            // already been shut down.
            //
            if (_instance == null)
            {
                return;
            }

            adapters = new List<Ice.ObjectAdapterI>(_adapters);

            _instance = null;
            _communicator = null;

            System.Threading.Monitor.PulseAll(this);
        }

        //
        // Deactivate outside the thread synchronization, to avoid
        // deadlocks.
        //
        foreach (Ice.ObjectAdapter adapter in adapters)
        {
            adapter.deactivate();
        }
    }

    public void waitForShutdown()
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            //
            // First we wait for the shutdown of the factory itself.
            //
            while (_instance != null)
            {
                System.Threading.Monitor.Wait(this);
            }

            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        //
        // Now we wait for deactivation of each object adapter.
        //
        foreach (Ice.ObjectAdapter adapter in adapters)
        {
            adapter.waitForDeactivate();
        }
    }

    public bool isShutdown()
    {
        lock (this)
        {
            return _instance == null;
        }
    }

    public void destroy()
    {
        //
        // First wait for shutdown to finish.
        //
        waitForShutdown();

        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        foreach (Ice.ObjectAdapter adapter in adapters)
        {
            adapter.destroy();
        }

        lock (this)
        {
            _adapters.Clear();
        }
    }

    public void
    updateConnectionObservers()
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        foreach (Ice.ObjectAdapterI adapter in adapters)
        {
            adapter.updateConnectionObservers();
        }
    }

    public void
    updateThreadObservers()
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        foreach (Ice.ObjectAdapterI adapter in adapters)
        {
            adapter.updateThreadObservers();
        }
    }

    public Ice.ObjectAdapter createObjectAdapter(
        string name,
        Ice.RouterPrx router,
        SslServerAuthenticationOptions serverAuthenticationOptions)
    {
        lock (this)
        {
            if (_instance == null)
            {
                throw new Ice.CommunicatorDestroyedException();
            }

            if (name.Length > 0)
            {
                if (_adapterNamesInUse.Contains(name))
                {
                    Ice.AlreadyRegisteredException ex = new Ice.AlreadyRegisteredException();
                    ex.kindOfObject = "object adapter";
                    ex.id = name;
                    throw ex;
                }
                _adapterNamesInUse.Add(name);
            }
        }

        //
        // Must be called outside the synchronization since initialize can make client invocations
        // on the router if it's set.
        //
        Ice.ObjectAdapterI adapter = null;
        try
        {
            if (name.Length == 0)
            {
                adapter = new Ice.ObjectAdapterI(
                    _instance,
                    _communicator,
                    this,
                    System.Guid.NewGuid().ToString(),
                    null,
                    true,
                    serverAuthenticationOptions);
            }
            else
            {
                adapter = new Ice.ObjectAdapterI(
                    _instance,
                    _communicator,
                    this,
                    name,
                    router,
                    false,
                    serverAuthenticationOptions);
            }

            lock (this)
            {
                if (_instance == null)
                {
                    throw new Ice.CommunicatorDestroyedException();
                }
                _adapters.Add(adapter);
            }
        }
        catch (Ice.CommunicatorDestroyedException)
        {
            if (adapter != null)
            {
                adapter.destroy();
            }
            throw;
        }
        catch (Ice.LocalException)
        {
            if (name.Length > 0)
            {
                lock (this)
                {
                    _adapterNamesInUse.Remove(name);
                }
            }
            throw;
        }

        return adapter;
    }

    public Ice.ObjectAdapter findObjectAdapter(Ice.ObjectPrx proxy)
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            if (_instance == null)
            {
                return null;
            }

            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        foreach (Ice.ObjectAdapterI adapter in adapters)
        {
            try
            {
                if (adapter.isLocal(proxy))
                {
                    return adapter;
                }
            }
            catch (Ice.ObjectAdapterDeactivatedException)
            {
                // Ignore.
            }
        }

        return null;
    }

    public void removeObjectAdapter(Ice.ObjectAdapterI adapter)
    {
        lock (this)
        {
            if (_instance == null)
            {
                return;
            }

            _adapters.Remove(adapter);
            _adapterNamesInUse.Remove(adapter.getName());
        }
    }

    public void flushAsyncBatchRequests(Ice.CompressBatch compressBatch, CommunicatorFlushBatchAsync outAsync)
    {
        List<Ice.ObjectAdapterI> adapters;
        lock (this)
        {
            adapters = new List<Ice.ObjectAdapterI>(_adapters);
        }

        foreach (Ice.ObjectAdapterI adapter in adapters)
        {
            adapter.flushAsyncBatchRequests(compressBatch, outAsync);
        }
    }

    //
    // Only for use by Instance.
    //
    internal ObjectAdapterFactory(Instance instance, Ice.Communicator communicator)
    {
        _instance = instance;
        _communicator = communicator;
        _adapterNamesInUse = new HashSet<string>();
        _adapters = new List<Ice.ObjectAdapterI>();
    }

    private Instance _instance;
    private Ice.Communicator _communicator;
    private HashSet<string> _adapterNamesInUse;
    private List<Ice.ObjectAdapterI> _adapters;
}

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/ObjectAdapterFactory.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package com.zeroc.IceInternal;

import com.zeroc.Ice.ObjectAdapter;
import com.zeroc.Ice.SSL.SSLEngineFactory;

public final class ObjectAdapterFactory {
    public void shutdown() {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            //
            // Ignore shutdown requests if the object adapter factory has
            // already been shut down.
            //
            if (_instance == null) {
                return;
            }

            adapters = new java.util.LinkedList<>(_adapters);
        }

        //
        // Deactivate outside the thread synchronization, to avoid
        // deadlocks.
        //
        for (ObjectAdapter adapter : adapters) {
            adapter.deactivate();
        }

        synchronized (this) {
            _instance = null;
            _communicator = null;
            notifyAll();
        }
    }

    public void waitForShutdown() {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            //
            // First we wait for the shutdown of the factory itself.
            //
            while (_instance != null) {
                try {
                    wait();
                } catch (InterruptedException ex) {
                    throw new com.zeroc.Ice.OperationInterruptedException();
                }
            }

            adapters = new java.util.LinkedList<>(_adapters);
        }

        //
        // Now we wait for deactivation of each object adapter.
        //
        for (ObjectAdapter adapter : adapters) {
            adapter.waitForDeactivate();
        }
    }

    public synchronized boolean isShutdown() {
        return _instance == null;
    }

    public void destroy() {
        //
        // First wait for shutdown to finish.
        //
        waitForShutdown();

        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            adapters = new java.util.LinkedList<>(_adapters);
        }

        for (ObjectAdapter adapter : adapters) {
            adapter.destroy();
        }

        synchronized (this) {
            _adapters.clear();
        }
    }

    public void updateConnectionObservers() {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            adapters = new java.util.LinkedList<>(_adapters);
        }

        for (ObjectAdapter adapter : adapters) {
            adapter.updateConnectionObservers();
        }
    }

    public void updateThreadObservers() {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            adapters = new java.util.LinkedList<>(_adapters);
        }

        for (ObjectAdapter adapter : adapters) {
            adapter.updateThreadObservers();
        }
    }

    public ObjectAdapter createObjectAdapter(
            String name, com.zeroc.Ice.RouterPrx router, SSLEngineFactory sslEngineFactory) {
        if (Thread.interrupted()) {
            throw new com.zeroc.Ice.OperationInterruptedException();
        }

        synchronized (this) {
            if (_instance == null) {
                throw new com.zeroc.Ice.CommunicatorDestroyedException();
            }

            if (!name.isEmpty()) {
                if (_adapterNamesInUse.contains(name)) {
                    throw new com.zeroc.Ice.AlreadyRegisteredException("object adapter", name);
                }
                _adapterNamesInUse.add(name);
            }
        }

        //
        // Must be called outside the synchronization since initialize can make client invocations
        // on the router if it's set.
        //
        com.zeroc.Ice.ObjectAdapter adapter = null;
        try {
            if (name.isEmpty()) {
                String uuid = java.util.UUID.randomUUID().toString();
                adapter =
                        new com.zeroc.Ice.ObjectAdapter(
                                _instance, _communicator, this, uuid, null, true, sslEngineFactory);
            } else {
                adapter =
                        new com.zeroc.Ice.ObjectAdapter(
                                _instance,
                                _communicator,
                                this,
                                name,
                                router,
                                false,
                                sslEngineFactory);
            }

            synchronized (this) {
                if (_instance == null) {
                    throw new com.zeroc.Ice.CommunicatorDestroyedException();
                }
                _adapters.add(adapter);
            }
        } catch (com.zeroc.Ice.CommunicatorDestroyedException ex) {
            if (adapter != null) {
                adapter.destroy();
            }
            throw ex;
        } catch (com.zeroc.Ice.LocalException ex) {
            if (!name.isEmpty()) {
                synchronized (this) {
                    _adapterNamesInUse.remove(name);
                }
            }
            throw ex;
        }

        return adapter;
    }

    public ObjectAdapter findObjectAdapter(com.zeroc.IceInternal.Reference ref) {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            if (_instance == null) {
                return null;
            }

            adapters = new java.util.LinkedList<>(_adapters);
        }

        for (ObjectAdapter adapter : adapters) {
            try {
                if (adapter.isLocal(ref)) {
                    return adapter;
                }
            } catch (com.zeroc.Ice.ObjectAdapterDeactivatedException ex) {
                // Ignore.
            }
        }

        return null;
    }

    public synchronized void removeObjectAdapter(ObjectAdapter adapter) {
        if (_instance == null) {
            return;
        }

        _adapters.remove(adapter);
        _adapterNamesInUse.remove(adapter.getName());
    }

    public void flushAsyncBatchRequests(
            com.zeroc.Ice.CompressBatch compressBatch, CommunicatorFlushBatch outAsync) {
        java.util.List<ObjectAdapter> adapters;
        synchronized (this) {
            adapters = new java.util.LinkedList<>(_adapters);
        }

        for (ObjectAdapter adapter : adapters) {
            adapter.flushAsyncBatchRequests(compressBatch, outAsync);
        }
    }

    //
    // Only for use by Instance.
    //
    ObjectAdapterFactory(Instance instance, com.zeroc.Ice.Communicator communicator) {
        _instance = instance;
        _communicator = communicator;
    }

    @SuppressWarnings("deprecation")
    @Override
    protected synchronized void finalize() throws Throwable {
        try {
            com.zeroc.IceUtilInternal.Assert.FinalizerAssert(_instance == null);
            com.zeroc.IceUtilInternal.Assert.FinalizerAssert(_communicator == null);
            com.zeroc.IceUtilInternal.Assert.FinalizerAssert(_adapters.isEmpty());
        } catch (java.lang.Exception ex) {
        } finally {
            super.finalize();
        }
    }

    private Instance _instance;
    private com.zeroc.Ice.Communicator _communicator;
    private java.util.Set<String> _adapterNamesInUse = new java.util.HashSet<>();
    private java.util.List<ObjectAdapter> _adapters = new java.util.LinkedList<>();
}
